File: shared/api-client/test/metric_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:capecho_api/capecho_api.dart';
import 'package:test/test.dart';

// The Dart metric contract must equal the committed fixture, which the TS validator also asserts
// against (backend/test/metric-contract.test.ts) — so a Dart↔TS drift fails CI (ENG-3 posture, same
// as the normalization golden vectors). cwd-tolerant fixture read (mirrors the normalization parity test).

Map<String, dynamic> _readFixture() {
  const candidates = <String>[
    'fixtures/metric-events-contract.json', // cwd = shared/api-client
    '../api-client/fixtures/metric-events-contract.json', // cwd = shared
    'shared/api-client/fixtures/metric-events-contract.json', // cwd = repo root
  ];
  for (final p in candidates) {
    final f = File(p);
    if (f.existsSync()) return jsonDecode(f.readAsStringSync()) as Map<String, dynamic>;
  }
  throw StateError('metric-events-contract.json not found (cwd=${Directory.current.path})');
}

void main() {
  final fx = _readFixture();

  test('the Dart metric contract matches the committed fixture (Dart↔TS parity anchor)', () {
    expect(kMetricContractVersion, fx['version']);
    expect(kMetricMaxDurationMs, fx['maxDurationMs']);

    final events = (fx['events'] as Map).cast<String, dynamic>();
    expect(kMetricEventContract.keys.toSet(), events.keys.toSet(), reason: 'event-type set');

    for (final entry in events.entries) {
      final type = entry.key;
      final spec = (entry.value as Map).cast<String, dynamic>();
      final dartSpec = kMetricEventContract[type]!;
      expect(dartSpec.needsClientRowId, spec['needsClientRowId'], reason: '$type.needsClientRowId');

      final fields = (spec['fields'] as Map).cast<String, dynamic>();
      expect(dartSpec.fields.keys.toSet(), fields.keys.toSet(), reason: '$type field set');
      for (final f in fields.entries) {
        final fj = (f.value as Map).cast<String, dynamic>();
        final df = dartSpec.fields[f.key]!;
        expect(df.type, fj['type'], reason: '$type.${f.key}.type');
        if (fj['type'] == 'int') {
          expect(df.min, fj['min'], reason: '$type.${f.key}.min');
          expect(df.max, fj['max'], reason: '$type.${f.key}.max');
        } else if (fj['type'] == 'enum') {
          expect(df.values, (fj['values'] as List).cast<String>(), reason: '$type.${f.key}.values');
        }
      }
    }
  });

  test('every MetricEventType enum value has a contract entry and vice versa', () {
    expect(MetricEventType.values.map((e) => e.wire).toSet(), kMetricEventContract.keys.toSet());
  });
}

File: backend/test/metric-contract.test.ts
import { test, expect } from "bun:test";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { METRIC_CONTRACT, METRIC_CONTRACT_VERSION, MAX_DURATION_MS } from "../src/metrics.ts";

// The TS validator must equal the committed fixture, which the Dart contract also asserts against
// (shared/api-client/test/metric_contract_test.dart) — so a Dart↔TS drift fails CI (ENG-3 posture).
const fixture = JSON.parse(
  readFileSync(fileURLToPath(new URL("../../shared/api-client/fixtures/metric-events-contract.json", import.meta.url)), "utf8"),
) as {
  version: number;
  maxDurationMs: number;
  events: Record<string, { needsClientRowId: boolean; fields: Record<string, { type: string; min?: number; max?: number; values?: string[] }> }>;
};

test("the TS METRIC_CONTRACT matches the committed fixture (Dart↔TS parity anchor)", () => {
  expect(METRIC_CONTRACT_VERSION).toBe(fixture.version);
  expect(MAX_DURATION_MS).toBe(fixture.maxDurationMs);
  expect(new Set(Object.keys(METRIC_CONTRACT))).toEqual(new Set(Object.keys(fixture.events)));

  for (const [type, spec] of Object.entries(fixture.events)) {
    const ts = METRIC_CONTRACT[type as keyof typeof METRIC_CONTRACT] as { needsClientRowId: boolean; fields: Record<string, { type: string; min?: number; max?: number; values?: readonly string[] }> };
    expect(ts.needsClientRowId).toBe(spec.needsClientRowId);
    expect(new Set(Object.keys(ts.fields))).toEqual(new Set(Object.keys(spec.fields)));
    for (const [fname, fj] of Object.entries(spec.fields)) {
      const tf = ts.fields[fname]!;
      expect(tf.type).toBe(fj.type);
      if (fj.type === "int") {
        expect(tf.min).toBe(fj.min);
        expect(tf.max).toBe(fj.max);
      } else if (fj.type === "enum") {
        expect([...(tf.values ?? [])]).toEqual(fj.values ?? []);
      }
    }
  }
});
